### ISSUE TITLE
Examples

### ISSUE THREAD

**Issue by svenknobloch** [phaazon/luminance-rs#38]
Hi! I was wondering if there are any examples or tests that are publicly available. I understand most of the functionality of the library but am having some trouble putting it all together, which one or two examples would really help with. Even just a quick rundown of how to render out a simple triangle would be quite helpful.

**Comment by phaazon** [comment]
Yeah! Two demos I wrote! You can have a look at [Céleri Rémoulade](https://github.com/phaazon/celeri-remoulade) and [Outline Invitation](https://github.com/phaazon/outline-2017-invitro). I wrote an engine for that, [spectra](https://github.com/phaazon/spectra), which uses luminance.

On a general note, you’re right, I should provide more example. I’ll try to do that today and tomorrow. The main idea is:

- create a context – you can use `luminance-glfw` or `luminance-glutin` (though the latest is not stable enough, I need to fix it)
- create a pipeline with a framebuffer (use the default one with `Framebuffer::default`), a shader program and a tess object (it contains the vertices). You can alternatively use an attributeless render if you don’t want to allocate a buffer on the GPU
- run the pipeline

You have a better example [here](https://gist.github.com/phaazon/8979aef6d1f20d6034466762f66b4e60).

I should definitely add those to the crate. How did you get into luminance by the way? :)

**Comment by svenknobloch** [comment]
Sweet, I'll check those out! Got a little confused on how to render with the pipeline since the functionality changed from the old `pipeline.run()` to `pipeline.enter()` and how exactly to go about using the shader commands. 

Discovered luminance on [arewegameyet](http://arewegameyet.com) and found it much simpler than gfx. Playing around with some graphics stuff to try and write a game :P

**Comment by phaazon** [comment]
Well, the gate stuff is because I came to the realisation (with Céleri Rémoulade) that expecting a slice (`&[ShadingCommand]` for instance) was too constraining and caused some data structure to be impossible to render. For instance, I have a lot of shared resources (`Rc<RefCell<_>>`) that I need to borrow from. Once I’ve borrowed them, I need to keep the borrowed object around until the actual render is performed. This was not possible before introducing the gate concept. The idea is that when you do something like `pipeline.enter(_)`, you pass in a continuation that accepts an object that will live until the pipeline dies. This enables the use of `Iterator` types and doesn’t require dynamic allocation in the rendering loop anymore.

I guess it’s a bit confusing and that I should write about it, indeed. Thanks for your feedback about arewegameyet, I didn’t even know my framework was listed there!

**Comment by svenknobloch** [comment]
Another quick question. I've been looking through the sample code you provided and drew a triangle (yay!) but I am still unsure as to what the `RawBuffer` and `RawTexture` slices are for in the pipeline. What exactly should I pass to those?

**Comment by phaazon** [comment]
Yeah, this part is a bit unstable, I need to refactor it a bit.

Those two slices are there when you want to bind a uniform buffer or textures. If you don’t have any, just pass empty slices. Those are then “indexed“ via uniforms (respectively, `Uniform<Binding>` and `Uniform<Unit>`).

**Comment by svenknobloch** [comment]
How would I map the uniforms/textures to the correct shader variable? When I pass a buffer filled with uniform data, it isn't being used 🙁

**Comment by phaazon** [comment]
If you have a buffer, let’s say `Buffer<Foo>`, you can get use a reference on it (it’ll be dereferenced to `RawBuffer`) and put that in your slice. Then you need to represent the buffer as a uniform to your shader with a `Uniform<Binding>` (0 in your case, since the buffer is at the index 0 in the slice).

This behavior is neat because it enables luminance to do aggressive optimization on the number of buffer bindings, but it’s not very intuitive and is prone to errors. As I said above, it might change in the future.

For instance, in my latest demo, I have a uniform representing a directional light ([here](https://github.com/phaazon/outline-2017-invitro/blob/master/src/forward_renderer.rs#L25)). Then, look at how I create a slice for that buffer [here](https://github.com/phaazon/outline-2017-invitro/blob/master/src/forward_renderer.rs#L73) and [here](https://github.com/phaazon/outline-2017-invitro/blob/master/src/forward_renderer.rs#L73).

So basically, you need:

- your buffer, that will be used in the slice so that luminance passes its handle to the GPU (in the slice)
- retain the index in that slice you put your buffer in
- use a uniform with a buffer binding with that index
- pass the uniform to the shader

I wish I had a better scheme to cope with that, but I haven’t found a way yet. For instance, [glium](https://crates.io/crates/glium) uses a total type for that – you need to pass all the values at once, if you need several, you have to duplicate your pipeline. The model for luminance is streaming and partial traversal (the borrow stuff I talked earlier).

**Comment by svenknobloch** [comment]
Can't seem to find a way to actually get a Uniform 😅. Uniform and UniformBuilder don't seem to have any constructors in the latest version. Also the ShadingGate method no longer takes an array of AlterUnis. That's what the binding buffer is for though right?

**Comment by phaazon** [comment]
Yeah I changed quite a bit of things. If you really use the latest version, the design is a bit cleaner to get uniforms. You have to use the `UniformBuilder::ask` function. It either gives you a uniform or a warning (inactive uniform, etc.).

**Comment by phaazon** [comment]
Also, the `unwrap_or_unbound` method is used so that you accumulate warnings in a `Vec` if something is wrong with a single uniform. Because you still need a uniform for the folded type (here it’s `ShaderToyUniforms`), you have two options:

- make the whole uniform interface fail, because you **need** all of its uniforms
- use an unbound uniform, that is a special kind of uniform that will just do nothing when you try to update it

**Comment by svenknobloch** [comment]
Ohhhhhh. I made the mistake of disregarding the second input from the shading gate 😬, my bad. I see where to get the uniforms now!

**Comment by svenknobloch** [comment]
It works! That was a silly mistake on my part. Thanks for clearing that up!

**Comment by phaazon** [comment]
Well, it’s also due to my mistake not to update the doc and have something clean. I’ll do something about it!

**Comment by svenknobloch** [comment]
Working on some texturing but I don't quite have it yet. From what I gathered, I need to assign the framebuffer a color slot, create a texture that matches that color slot, upload to the texture, bind a unit uniform to it and then pass the texture buffer to the render gate. Is there something I'm missing?

**Comment by phaazon** [comment]
If you need to texturize an object, all you need is a simple texture. You just bind it as a uniform and pass the texture buffer. You have nothing to do with the framebuffer – if you really need color slot(s), luminance creates the texture(s) for you.

**Comment by svenknobloch** [comment]
I'll leave that out then. Also which gate should the texture slice be passed to?

**Comment by phaazon** [comment]
Which one you want. The idea is to be able to texturize at every level in the tree. Bind as soon as you have it (the pipeline entry is a good point).

**Comment by svenknobloch** [comment]
Still not showing up 🙁.
This would be a valid texture correct?
```
let tex = Texture::<Flat, Dim2, RGBA8UI>::new([1, 1], 0, &sampler).unwrap();  
tex.upload(false, &[(255, 0, 0, 255)]);
```

**Comment by phaazon** [comment]
You’re using the [upload](https://docs.rs/luminance/0.21.2/luminance/texture/struct.Texture.html#method.upload) function. It expects the texels’s size to match. If you want to clear the texture, use the [clear](https://docs.rs/luminance/0.21.2/luminance/texture/struct.Texture.html#method.upload) function.

**Comment by svenknobloch** [comment]
That should make a 1x1 texture and then upload a single red pixel, no?

**Comment by phaazon** [comment]
Yeah, but I think 1x1 textures are forbidden by OpenGL. I should check that, actually, in luminance!

**Comment by phaazon** [comment]
I must integrate the example of [my latest blog entry about luminance](http://phaazon.blogspot.fr/2017/09/state-of-luminance.html). I’ll close this issue when it’s done.

**Comment by phaazon** [comment]
Ok, I think there’s enough examples for now. I’m closing this. I’ll add more example when I think about them, but for now, I guess it’s quite okay.
